app_dart/lib/src/model/firestore/pr_check_runs.dart

```diff
@@ -0,0 +1,111 @@
+  static PrCheckRuns fromDocument(Document prCheckRunsDoc) {
+    return PrCheckRuns()
+      ..fields = prCheckRunsDoc.fields!
+      ..name = prCheckRunsDoc.name!;
+  }
+
+  /// The json string of the pullrequest belonging to this document.
+  PullRequest get pullRequest => PullRequest.fromJson(json.decode(fields![kPullRequestField]!.stringValue!));
+
+  /// The head sha at the time this document was created for testing.
+  String get sha => fields![kShaField]!.stringValue!;
+
+  RepositorySlug get slug => RepositorySlug.fromJson(json.decode(fields![kSlugField]!.stringValue!));
+
+  /// Initializes a new document for the list of check_runs in Firestore so we can find it later.
+  ///
+  /// The list of tasks will be written as fields of a document with additional fields for tracking the total
+  /// number of tasks, remaining count. It is required to include [checkRunGuard] as a json encoded [CheckRun] as this
+  /// will be used to unlock any check runs blocking progress.
+  ///
+  /// Returns the created document or throws an error.
+  static Future<Document> initializeDocument({
+    required FirestoreService firestoreService,
+    required PullRequest pullRequest,
+    required List<CheckRun> checks,
+  }) async {
+    final logCrumb =
+        'initializeDocument(${pullRequest.repo?.slug().fullName}/${pullRequest.number}, ${checks.length} check runs)';
+
+    final fields = <String, Value>{
+      kPullRequestField: Value(stringValue: json.encode(pullRequest.toJson())),
+      kSlugField: Value(stringValue: json.encode(pullRequest.head!.repo!.slug().toJson())),
+      kShaField: Value(stringValue: pullRequest.head!.sha!),
+      for (final run in checks) '${run.id}': Value(stringValue: run.name),
+    };
+
+    final document = Document(fields: fields);
+
+    try {
+      // Calling createDocument multiple times for the same documentId will return a 409 - ALREADY_EXISTS error;
+      // this is good because it means we don't have to do any transactions.
+      // curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer <TOKEN>" "https://firestore.googleapis.com/v1beta1/projects/flutter-dashboard/databases/cocoon/documents/prCheckRuns" -d '{"fields": {"test": {"stringValue": "baz"}}}'
+      final databasesDocumentsResource = await firestoreService.documentResource();
+      final newDoc = await databasesDocumentsResource.createDocument(
+        document,
+        kDocumentParent,
+        kCollectionId,
+      );
+      log.info('$logCrumb: document created');
+      return newDoc;
+    } catch (e) {
+      log.warning('$logCrumb: failed to create document: $e');
+      rethrow;
+    }
+  }
+
+  /// Retrieve the [PullRequest] for a given [checkRun] or throw an error.
+  static Future<PullRequest> findDocumentFor(
+    FirestoreService firestoreService,
+    CheckRun checkRun,
+  ) async {
+    final filterMap = <String, Object>{
+      '${checkRun.id} =': checkRun.name!,
```

Comment:
I admit, I don't understand how this filter map works. This will look like `{ 'abc123': 'Linux host_debug' }` or something, and I'm not sure how that tells firestore to query for the check run with that ID, and what it has to do with the check run's name. I expected to see something like `{ 'id': 'abc123' }`.

Reply:
I have no clue what's going on here other than someone wrote a util on top of firestore's apis. I assume this works since its used elsewhere.  What I'm looking for is specificity - the id must exist as a field and the name must be the same as what it was created with (in case GitHub is lying about the uniqueness of check run ids). If we find a doc with that unique signature; then we're good to go with the PR.

I'm realizing now that firebase might not like this if we're creating single index keys https://firebase.google.com/docs/firestore/query-data/index-overview#single-field_indexes

I should flip this and have it as `"name": "id"` and it would mean the key space is ~150.

Reply:
To be clear; the util is parsing this map and making some newer filter thing that ultimiatly looks like "{field: fu, op: EQUALS, value: bar}".

```dart
  /// The [filterMap] follows format:
  ///   {
  ///     'fieldInt =': 1,
  ///     'fieldString =': 'string',
  ///     'fieldBool =': true,
  ///   }
  /// Note
  ///   1. the space in the key, which will be used to retrieve the field name and operator.
  ///   2. the value could be any type, like int, string, bool, etc.
  ```